### PR TITLE
Add types to all of parameter

### PR DESCRIPTION
**Pull request by jenshnielsen**
There are a number of less than idea issues here that makes it tricky and requires workarounds

### CODE REVIEW

File: qcodes/instrument/parameter.py [hunk 1]
```diff
@@ -1717,17 +1732,20 @@ def __call__(self) -> Any:
         return self.get()
 
 
-def combine(*parameters, name, label=None, unit=None, units=None,
-            aggregator=None):
+def combine(*parameters: 'Parameter',
+            name: str,
+            label: Optional[str] = None,
+            unit: str = None,
+            units: str = None,
```
Comment (jenshnielsen) [inline qcodes/instrument/parameter.py]: These should probably be optional

File: qcodes/instrument/parameter.py [hunk 2]
```diff
@@ -2042,12 +2094,13 @@ def __init__(self,
...
+        if division is not None:
             self.role = ScaledParameter.Role.DIVISION
-            self._multiplier = division
-        elif is_amplifier:
+            # mypy issue 3004
+            self._multiplier = division  # type: ignore
+        elif gain is not None:
             self.role = ScaledParameter.Role.GAIN
-            self._multiplier = gain
+            self._multiplier = gain  # type: ignore
```
Comment (jenshnielsen) [inline qcodes/instrument/parameter.py]: Needs error-code

File: qcodes/instrument/parameter.py [hunk 3]
```diff
@@ -2000,11 +2052,11 @@ class Role(enum.Enum):
 
     def __init__(self,
                  output: Parameter,
-                 division: Union[int, float, Parameter] = None,
-                 gain: Union[int, float, Parameter] = None,
-                 name: str=None,
-                 label: str=None,
-                 unit: str=None) -> None:
+                 division: Optional[Union[int, float, Parameter]] = None,
+                 gain: Optional[Union[int, float, Parameter]] = None,
```
Comment (jenshnielsen) [inline qcodes/instrument/parameter.py]: Use number type here

File: qcodes/instrument/parameter.py [hunk 4]
```diff
@@ -2060,11 +2113,13 @@ def __init__(self,
     # Internal handling of the multiplier
     # can be either a Parameter or a scalar
     @property
-    def _multiplier(self):
+    def _multiplier(self) -> Parameter:
+        if self._multiplier_parameter is None:
+            raise RuntimeError
```
Comment (jenshnielsen) [inline qcodes/instrument/parameter.py]: Add Text

File: qcodes/instrument/parameter.py [hunk 5]
```diff
@@ -2075,40 +2130,48 @@ def _multiplier(self, multiplier: Union[int, float, Parameter]):
 
     # Division of the scaler
     @property
-    def division(self):
+    def division(self) -> float:  # type: ignore
```
Comment (jenshnielsen) [inline qcodes/instrument/parameter.py]: Add error type

File: qcodes/instrument/parameter.py [hunk 6]
```diff
@@ -2042,12 +2094,13 @@ def __init__(self,
         if not xor(is_divider, is_amplifier):
             raise ValueError('Provide only division OR gain')
 
-        if is_divider:
+        if division is not None:
             self.role = ScaledParameter.Role.DIVISION
-            self._multiplier = division
-        elif is_amplifier:
+            # mypy issue 3004
```
Comment (jenshnielsen) [inline qcodes/instrument/parameter.py]: Flesh out the comment to reflect why mypy issue 3004 means that this cannot be done probably

File: qcodes/instrument/parameter.py [hunk 7]
```diff
@@ -2075,40 +2130,48 @@ def _multiplier(self, multiplier: Union[int, float, Parameter]):
...
+        self._multiplier = gain  # type: ignore
 
     # Getter and setter for the real value
     def get_raw(self) -> Union[int, float]:
         """
         Returns:
             value at which was set at the sample
         """
+        wrapped_value = cast(Union[float, int], self._wrapped_parameter())
+        multiplier = cast(Union[float, int], self._multiplier())
```
Comment (jenshnielsen) [inline qcodes/instrument/parameter.py]: Number

File: qcodes/instrument/parameter.py [hunk 8]
```diff
@@ -2075,40 +2130,48 @@ def _multiplier(self, multiplier: Union[int, float, Parameter]):
...
         if self.role == ScaledParameter.Role.DIVISION:
-            return self._multiplier()
+            return value
         elif self.role == ScaledParameter.Role.GAIN:
-            return 1 / self._multiplier()
+            return 1 / value
 
     @division.setter
-    def division(self, division: Union[int, float, Parameter]):
+    def division(self, division: Union[int, float, Parameter]) -> None:
```
Comment (jenshnielsen) [inline qcodes/instrument/parameter.py]: Number

File: qcodes/instrument/parameter.py [hunk 9]
```diff
@@ -125,13 +125,15 @@ def __init__(self, parameter: "Parameter", value: Any):
...
-    def __enter__(self):
+    def __enter__(self) -> None:
         if self._value_is_changing:
-            self._parameter.set(self._value)
+            self._parameter.set(self._value)  # type: ignore[has-type]
 
-    def __exit__(self, typ, value, traceback):
+    def __exit__(self, typ,  # type: ignore[no-untyped-def]
+                 value,
+                 traceback) -> None:
```
[the remark this answers is not in the document]
Comment (astafan8) [reply qcodes/instrument/parameter.py]: yeah....

File: qcodes/instrument/parameter.py [hunk 10]
```diff
@@ -316,7 +317,7 @@ def __init__(self, name: str,
         self._validate_on_get = False
 
     @abstractmethod
-    def get_raw(self):
+    def get_raw(self) -> Any:
```
Comment (astafan8) [inline qcodes/instrument/parameter.py]: I added `ParamRawDataType` in https://github.com/QCoDeS/Qcodes/pull/1757 :)
```suggestion
    def get_raw(self) -> ParamRawDataType:
```

File: qcodes/instrument/parameter.py [hunk 11]
```diff
@@ -329,7 +330,7 @@ def get_raw(self):
         raise NotImplementedError
 
     @abstractmethod
-    def set_raw(self, value):
+    def set_raw(self, value: Any) -> None:
```
Comment (astafan8) [inline qcodes/instrument/parameter.py]: I added `ParamRawDataType` in https://github.com/QCoDeS/Qcodes/pull/1757 :)
```suggestion
    def set_raw(self, value: ParamRawDataType) -> None:
```

File: qcodes/instrument/parameter.py [hunk 12]
```diff
@@ -371,8 +372,8 @@ def snapshot_base(self, update: bool = True,
                       ) -> Dict:
         """
         State of the parameter as a JSON-compatible dict (everything that
-        the custom JSON encoder class :class:'qcodes.utils.helpers.NumpyJSONEncoder'
-        supports).
+        the custom JSON encoder class
+        :class:'qcodes.utils.helpers.NumpyJSONEncoder' supports).
```
Comment (astafan8) [inline qcodes/instrument/parameter.py]: unrelated to this PR, but:
```suggestion
        :class:`qcodes.utils.helpers.NumpyJSONEncoder` supports).
```

File: qcodes/instrument/parameter.py [hunk 13]
```diff
@@ -600,9 +602,8 @@ def get_ramp_values(self, value: Union[float, int, Sized],
                 # isn't, even though it's valid.
                 # probably MultiType with a mix of numeric and non-numeric types
                 # just set the endpoint and move on
-                log.warning(
-                    'cannot sweep %s from %r to %r - jumping.',
-                                self.name, start_value, value)
+                log.warning(f'cannot sweep {self.name} from {start_value} '
+                            f'to {value} - jumping.')
```
Comment (astafan8) [inline qcodes/instrument/parameter.py]:
```suggestion
                            f'to {value:!r} - jumping.')
```

File: qcodes/instrument/parameter.py [hunk 14]
```diff
@@ -600,9 +602,8 @@ def get_ramp_values(self, value: Union[float, int, Sized],
                 # isn't, even though it's valid.
                 # probably MultiType with a mix of numeric and non-numeric types
                 # just set the endpoint and move on
-                log.warning(
-                    'cannot sweep %s from %r to %r - jumping.',
-                                self.name, start_value, value)
+                log.warning(f'cannot sweep {self.name} from {start_value} '
```
Comment (astafan8) [inline qcodes/instrument/parameter.py]:
```suggestion
                log.warning(f'cannot sweep {self.name} from {start_value:!r} '
```

File: qcodes/instrument/parameter.py [hunk 15]
```diff
@@ -799,7 +801,7 @@ def root_instrument(self) -> Optional['InstrumentBase']:
         else:
             return None
 
-    def set_to(self, value):
+    def set_to(self, value: Any) -> _SetParamContext:
```
Comment (astafan8) [inline qcodes/instrument/parameter.py]: perhaps?
```suggestion
    def set_to(self, value: ParamDataType) -> _SetParamContext:
```

File: qcodes/instrument/parameter.py [hunk 16]
```diff
@@ -1014,32 +1020,34 @@ def __init__(self, name: str,
...
-    def __getitem__(self, keys):
+    def __getitem__(self, keys: Any) -> 'SweepFixedValues':
         """
         Slice a Parameter to get a SweepValues object
         to iterate over during a sweep
         """
         return SweepFixedValues(self, keys)
 
-    def increment(self, value):
+    def increment(self, value: Any) -> None:
```
Comment (astafan8) [inline qcodes/instrument/parameter.py]:
```suggestion
    def increment(self, value: ParamDataType) -> None:
```

File: qcodes/instrument/parameter.py [hunk 17]
```diff
@@ -1717,17 +1729,20 @@ def __call__(self) -> Any:
...
-        *parameters (qcodes.instrument.parameter.Parameter): The parameters to
-            combine.
-        name (str): The name of the paramter.
-        label (Optional[str]): The label of the combined parameter.
-        unit (Optional[str]): the unit of the combined parameter.
+        *parameters: The parameters to combine.
+        name: The name of the paramter.
+        label: The label of the combined parameter.
+        unit: the unit of the combined parameter.
         aggregator (Optional[Callable[list[Any]]]): a function to aggregate
```
Comment (astafan8) [inline qcodes/instrument/parameter.py]:
```suggestion
        aggregator: a function to aggregate
```

File: qcodes/instrument/parameter.py [hunk 18]
```diff
@@ -2075,40 +2132,48 @@ def _multiplier(self, multiplier: Union[int, float, Parameter]):
 
     # Division of the scaler
     @property
-    def division(self):
+    def division(self) -> float:  # type: ignore[return]
```
Comment (astafan8) [inline qcodes/instrument/parameter.py]: why ignore return if the type is specified via casting? or am i misunderstanding something?

File: qcodes/instrument/parameter.py [hunk 19]
```diff
@@ -2075,40 +2132,48 @@ def _multiplier(self, multiplier: Union[int, float, Parameter]):
...
+
         if self.role == ScaledParameter.Role.GAIN:
-            value = self._wrapped_parameter() * self._multiplier()
+            value = wrapped_value * multiplier
         elif self.role == ScaledParameter.Role.DIVISION:
-            value = self._wrapped_parameter() / self._multiplier()
+            value = wrapped_value / multiplier
+        else:
+            raise RuntimeError(f"ScaledParameter must be either a"
+                               f"Multiplier or Divisor got {self.role}")
```
Comment (astafan8) [inline qcodes/instrument/parameter.py]:
```suggestion
                               f"Multiplier or Divisor; got {self.role}")
```

File: qcodes/instrument/parameter.py [hunk 20]
```diff
@@ -2120,25 +2185,26 @@ def wrapped_parameter(self) -> Parameter:
...
+            instrument_value = value / multiplier_value
         elif self.role == ScaledParameter.Role.DIVISION:
-            instrument_value = value * self._multiplier()
-
-        # don't leak unknow type
-        instrument_value = cast(Union[int, float], instrument_value)
+            instrument_value = value * multiplier_value
+        else:
+            raise RuntimeError(f"ScaledParameter must be either a"
+                               f"Multiplier or Divisor got {self.role}")
```
Comment (astafan8) [inline qcodes/instrument/parameter.py]:
```suggestion
                               f"Multiplier or Divisor; got {self.role}")
```

File: qcodes/instrument/parameter.py (hunk 18 above)
Comment (jenshnielsen) [reply qcodes/instrument/parameter.py]: Due to the mypy bug mentioned above. Remember that unused ignores is an error, so you can be sure this is needed anyway

Comment (jenshnielsen) [reply qcodes/instrument/parameter.py]: That is if mypy ever resolves the bug this will be an error and all to do is to remove this ignore comment

**Pull request merged**